core/citrus-api/src/main/java/com/consol/citrus/variable/VariableExpressionSegmentMatcher.java

```diff
@@ -0,0 +1,187 @@
+ *       <li>var1[1].var2[2]</li>
+ *       <li>var1[1].var2[2].var3</li>
+ *       <li>var1.jsonPath($.person.name)</li>
+ *       <li>var1[1].jsonPath($.person.name)</li>
+ *       <li>var1.xpath(//title[@lang='en'])</li>
+ *       <li>var1[1].xpath(//title[@lang='en'])</li>
+ *   </ul>
+ *   <br>
+ */
+public class VariableExpressionSegmentMatcher {
+
+    /**
+     * Pattern to parse a variable expression
+     */
+    private static final Pattern VAR_PATH_PATTERN = Pattern.compile("(xpath\\((.*)\\)$)|(jsonPath\\((\\$[.\\[].*)\\)$)|(([^\\[\\].]+)(\\[([0-9])])?)(\\.|$)");
+
+    /**
+     * The regex group index for the full xpath segment
+     */
+    private static final int XPATH_SEGMENT_GROUP = 1;
+
+    /**
+     * The regex group index for the xpath path
+     */
+    private static final int XPATH_GROUP = 2;
+
+    /**
+     * The regex group index for the full jsonPath segment
+     */
+    private static final int JSONPATH_SEGMENT_GROUP = 3;
+
+    /**
+     * The regex group index for the jsonPath part
+     */
+    private static final int JSON_PATH_GROUP = 4;
+
+    /**
+     * The regex group index for the full variable/property segment incl. index
+     */
+    private static final int VAR_PROP_SEGMENT_GROUP = 5;
+
+    /**
+     * The regex group index for the name of the variable/property
+     */
+    private static final int VAR_PROP_NAME_GROUP = 6;
+
+    /**
+     * The regex group index for the full name/index expression
+     */
+    private static final int NAME_INDEX_GROUP = 7;
+
+    /**
+     * The regex group index for the index when accessing array elements
+     */
+    private static final int INDEX_GROUP = 8;
+
+    /**
+     * The variable expression the matcher is working on
+     */
+    private final String variableExpression;
+
+    /**
+     * The matcher that performs the actual matching
+     */
+    private final Matcher matcher;
+
+    /**
+     * The total number of segments in the variableExpression
+     */
+    private final int totalSegmentCount;
+
+    /**
+     * The current expression the matcher has matched
+     */
+    private String currentSegmentExpression;
+
+    /**
+     * The current segment expression index. A value of -1 indicates "no index".
+     */
+    private int currentSegmentIndex = -1;
+
+    public VariableExpressionSegmentMatcher(String variableExpression) {
+        this.variableExpression = variableExpression;
+        totalSegmentCount =  totalSegmentCount(variableExpression);
+        matcher = VAR_PATH_PATTERN.matcher(variableExpression);
+    }
+
+    /**
+     * Determine the total number of segments in the variable expression
+     * @param variableExpression
+     * @return
+     */
+    private int totalSegmentCount(String variableExpression) {
+        Matcher matcher = VAR_PATH_PATTERN.matcher(variableExpression);
+        int segmentCounter = 0;
+        while (matcher.find()) {
+            segmentCounter++;
+        }
+        return segmentCounter;
```

Comment:
I think you can ask for `matcher.results().count()` or something.

Reply:
That's nice. I was actually looking for something like that. Obviously not eagerly enough.